docs/async-await.md

```diff
@@ -0,0 +1,168 @@
+# Proposal: Async-await support
+
+## Introduction
+
+With the introduction of [async/await][SE-0296] in Swift 5.5, it is now possible to write asynchronous code without the need for callbacks. 
+
+Language support for [`AsyncSequence`][SE-0298] also allows for writing functions that return values over time.
+
+We would like to explore how we could offer APIs that make use of these new language features to allow users to run HTTPRequest using these new idioms.
+
+This proposal describes what these APIs could look like and explores some of the potential usability concerns.
+
+## Proposed API additions
+
+### New `AsyncRequest` type
+
+The proposed new `AsyncRequest` shall be a simple swift structure.
+
+```swift
+struct AsyncRequest {
+  /// The requests url.
+  var url: String
+  
+  /// The request's HTTPMethod
+  var method: HTTPMethod
+  
+  /// The request's headers
+  var headers: HTTPHeaders
+
+  /// The request's body
+  var body: Body?
```

Comment:
@ktoso hmm, the above `HTTPEntity.*` list seems extremely HTTP/1.1 specific and even exposes some really bad things like "close delimited" which we don't need. Also a user shouldn't care if it's sent using chunked encoding or not, right? That's a choice the HTTP lib can make itself. If the user doesn't tell the number of bytes ahead of time, then in HTTP/1.1 it should use chunked encoding. If the user promises the size ahead of time, it can use content-length instead. It should never use EOF-framing (aka "close delimited").

- `Empty` seems like an (unnecessary?) specialisation of a body with a pre-announced length (of 0 bytes)
- `indefiniteLength` seems like an (unnecessary?) specialisation of a body without a pre-announced length

Reply:
I spoke with @ktoso offline that there might be performance optimization potential in here. @ktoso mentioned, that in akka they could optimize for cases where they received the response body in the same read as the response head. In those cases they would not be required to start a stream, but just copy the bytes out.

We might be able to achieve the same if we changed NIOHTTP1. However we should not expose this optimizations to the user with an API option. For this reason we should stay with the currently proposed API.

Reply:
@fabianfett you can do the same in `NIOHTTP1`. If you see `channelRead(.head(...))`, `channelRead(.body(...))`, `channelRead(.end(...))`, `channelReadComplete`, then you have everything in one go and you don't need to stream. If you get `channelReadComplete` before `channelRead(.end(...))` then you'd need to stream.

Reply:
I agree. We don't need to change `NIOHTTP1` in this case. And we even use `channelReadComplete` today. This would entail some internal changes though. But could be a great optimization.